File: sportix-backend/app/services/reel_service.py
from appwrite.query import Query as Q
from appwrite.id import ID
from app.core.appwrite import db, DB_ID
from app.core.config import settings
from app.utils.formatters import now_iso
from app.schemas.post import ReelCreate
from typing import Optional


async def get_feed(user_id: str, page: int, sport: Optional[str]) -> dict:
    queries = [Q.limit(20), Q.offset(page * 20), Q.order_desc("$createdAt")]
    if sport:
        queries.append(Q.equal("sport_tag", sport))
    return db.list_documents(DB_ID, settings.collection_reels, queries=queries)


async def create(user_id: str, payload: ReelCreate) -> dict:
    return db.create_document(
        DB_ID, settings.collection_reels, ID.unique(),
        data={
            "created_at": now_iso(),
            "author_id": user_id,
            "video_url": payload.video_url,
            "thumbnail_url": payload.thumbnail_url,
            "caption": payload.caption,
            "sport_tag": payload.sport_tag,
            "duration_seconds": payload.duration_seconds,
            "likes_count": 0,
            "views_count": 0,
        },
    )


async def get_by_id(reel_id: str, viewer_id: str) -> dict:
    doc = db.get_document(DB_ID, settings.collection_reels, reel_id)
    doc["is_liked"] = _check_liked(reel_id, viewer_id)
    return doc


async def get_by_user(user_id: str, page: int) -> dict:
    return db.list_documents(
        DB_ID, settings.collection_reels,
        queries=[Q.equal("author_id", user_id), Q.limit(20), Q.offset(page * 20), Q.order_desc("$createdAt")],
    )


async def delete(reel_id: str, user_id: str):
    doc = db.get_document(DB_ID, settings.collection_reels, reel_id)
    if doc.get("author_id") != user_id:
        raise PermissionError("You can only delete your own reels")
    db.delete_document(DB_ID, settings.collection_reels, reel_id)


async def toggle_like(reel_id: str, user_id: str) -> dict:
    existing = db.list_documents(
        DB_ID, settings.collection_reel_likes,
        queries=[Q.equal("reel_id", reel_id), Q.equal("user_id", user_id), Q.limit(1)],
    )
    if existing.get("documents"):
        db.delete_document(DB_ID, settings.collection_reel_likes, existing["documents"][0]["$id"])
        _bump_count(reel_id, "likes_count", -1)
        return {"liked": False}
    else:
        db.create_document(DB_ID, settings.collection_reel_likes, ID.unique(),
                           { "created_at": now_iso(),"reel_id": reel_id, "user_id": user_id})
        _bump_count(reel_id, "likes_count", 1)
        return {"liked": True}


def _check_liked(reel_id: str, user_id: str) -> bool:
    try:
        res = db.list_documents(
            DB_ID, settings.collection_reel_likes,
            queries=[Q.equal("reel_id", reel_id), Q.equal("user_id", user_id), Q.limit(1)],
        )
        return len(res.get("documents", [])) > 0
    except Exception:
        return False


def _bump_count(reel_id: str, field: str, delta: int):
    try:
        r = db.get_document(DB_ID, settings.collection_reels, reel_id)
        db.update_document(DB_ID, settings.collection_reels, reel_id,
                           {field: max(0, r.get(field, 0) + delta)})
    except Exception:
        pass

File: app-sportix/src/services/reelService.ts
/**
 * src/services/reelService.ts
 */
import { account, databases, DATABASE_ID, COLLECTIONS, ID, Query } from '../api/appwrite';
import { Reel } from '../types';
import { storageService } from './storageService';

function toReel(doc: any): Reel {
  return {
    $id:               doc.$id,
    author_id:         doc.author_id,
    author_name:       doc.author_name       ?? '',
    author_username:   doc.author_username   ?? '',
    author_avatar_url: doc.author_avatar_url ?? null,
    author_sport:      doc.author_sport      ?? '',
    video_url:         doc.video_url         ?? '',
    thumbnail_url:     doc.thumbnail_url     ?? null,
    caption:           doc.caption           ?? '',
    sport_tag:         doc.sport_tag         ?? '',
    likes_count:       doc.likes_count       ?? 0,
    comments_count:    doc.comments_count    ?? 0,
    views_count:       doc.views_count       ?? 0,
    $createdAt:        doc.$createdAt,
  };
}

export const reelService = {
  async getReels(offset = 0, limit = 10): Promise<Reel[]> {
    const res = await databases.listDocuments(DATABASE_ID, COLLECTIONS.REELS, [
      Query.orderDesc('$createdAt'),
      Query.limit(limit),
      Query.offset(offset),
    ]);
    return res.documents.map(toReel);
  },

  async toggleLike(reelId: string): Promise<boolean> {
    const raw = await account.get();
    const existing = await databases.listDocuments(DATABASE_ID, COLLECTIONS.REEL_LIKES, [
      Query.equal('reel_id', reelId),
      Query.equal('user_id', raw.$id),
      Query.limit(1),
    ]);

    const reel = await databases.getDocument(DATABASE_ID, COLLECTIONS.REELS, reelId);

    if (existing.total > 0) {
      await databases.deleteDocument(DATABASE_ID, COLLECTIONS.REEL_LIKES, existing.documents[0].$id);
      await databases.updateDocument(DATABASE_ID, COLLECTIONS.REELS, reelId, {
        likes_count: Math.max(0, ((reel as any).likes_count ?? 1) - 1),
      });
      return false;
    } else {
      await databases.createDocument(DATABASE_ID, COLLECTIONS.REEL_LIKES, ID.unique(), {
        reel_id:    reelId,
        user_id:    raw.$id,
        created_at: new Date().toISOString(),
      });
      await databases.updateDocument(DATABASE_ID, COLLECTIONS.REELS, reelId, {
        likes_count: ((reel as any).likes_count ?? 0) + 1,
      });
      return true;
    }
  },

  async isLiked(reelId: string): Promise<boolean> {
    try {
      const raw = await account.get();
      const res = await databases.listDocuments(DATABASE_ID, COLLECTIONS.REEL_LIKES, [
        Query.equal('reel_id', reelId),
        Query.equal('user_id', raw.$id),
        Query.limit(1),
      ]);
      return res.total > 0;
    } catch {
      return false;
    }
  },

  async uploadReel(data: {
    videoUri:   string;
    caption:    string;
    sport_tag?: string;
  }): Promise<Reel> {
    const raw = await account.get();
    const { profileService } = await import('./profileService');
    const profile = await profileService.getProfile(raw.$id);

    const { url } = await storageService.upload(data.videoUri, 'video/mp4', `reel_${Date.now()}.mp4`);

    const doc = await databases.createDocument(DATABASE_ID, COLLECTIONS.REELS, ID.unique(), {
      author_id:         raw.$id,
      author_name:       profile?.full_name  ?? raw.name,
      author_username:   profile?.username   ?? '',
      author_avatar_url: profile?.avatar_url ?? null,
      author_sport:      profile?.sport      ?? '',
      video_url:         url,
      thumbnail_url:     null,
      caption:           data.caption,
      sport_tag:         data.sport_tag ?? '',
      likes_count:       0,
      comments_count:    0,
      views_count:       0,
    });
    return toReel(doc);
  },
};
